### PR TITLE
Store a mask; Much smaller memory footprint

### PR DESCRIPTION
**Pull request by mrcslws**
Rather than storing a list of zero indices, store a zero mask.

With this change, the SparseWeightsBase no longer prescribes how the zeroes should be stored; the subclass gets to decide this. This enables subclasses to choose more efficient case-specific rezeroing techniques, for example when sparsity is structured (e.g. channel-to-channel).

**Memory usage of resnet50, before and after:**

Before:
- 672,979,068 bytes total
- 570,523,648 bytes of SparseWeights zero indices

After:
- 152,778,876 bytes total
- 50,323,456 bytes of SparseWeights masks

And the memory improvement gets considerably better if the sparsity is structured (e.g. channel to channel), enabling broadcasting.

### CODE REVIEW

File: nupic/torch/modules/sparse_weights.py [hunk 1]
```diff
@@ -136,33 +123,27 @@ class SparseWeights(SparseWeightsBase):
     """
 
     def __init__(self, module, weight_sparsity=None, sparsity=None):
-        assert isinstance(module, nn.Linear)
-        assert 0 < (weight_sparsity or sparsity) < 1
+        assert len(module.weight.shape) == 2, "Should resemble a nn.Linear"
+        assert 0 <= (weight_sparsity or sparsity) <= 1
```
Comment (lscheinkman) [inline nupic/torch/modules/sparse_weights.py:127]: This assert was placed to prevent the creation of `SparseWeight` layers with 0% or 100% sparsity. Both case should be considered to be a bug.
- `SparseWeight` with `sparsity=0` is not sparse, it similar to `torch.nn.Identity` but using more memory to hold the dense mask.
- `SparseWeight` with `sparsity=1` is not sparse, it is the same as replacing the layer with zeros.

These edge cases make sense in the `PrunableSparseWeights` context where the sparsity may be changed during training but not in the static network context where the sparsity is fixed at initialization.

See discussion in https://github.com/numenta/nupic.torch/pull/67

File: nupic/torch/modules/sparse_weights.py [hunk 2]
```diff
@@ -178,35 +159,28 @@ class SparseWeights2d(SparseWeightsBase):
     """
 
     def __init__(self, module, weight_sparsity=None, sparsity=None):
-        assert isinstance(module, nn.Conv2d)
-        assert 0 < (weight_sparsity or sparsity) < 1
+        assert len(module.weight.shape) == 4, "Should resemble a nn.Conv2d"
+        assert 0 <= (weight_sparsity or sparsity) <= 1
```
Comment (lscheinkman) [inline nupic/torch/modules/sparse_weights.py:163]: This assert was placed to prevent the creation of SparseWeight2d layers with 0% or 100% sparsity. Both cases should be considered to be a bug.
See comment on `SparseWeights` class

File: nupic/torch/modules/sparse_weights.py (hunk 1 above)
Comment (mrcslws) [reply nupic/torch/modules/sparse_weights.py:127]: Can I add a `allow_extremes=False` bool to the `__init__` that allows these values? The problem with having these asserts is that every subclass needs to be careful not to call `SparseWeightsBase.__init__`, which makes `super()` code really complicated. If subclasses have the option of setting `allow_extremes=True`, it would solve that problem.

File: nupic/torch/compatibility.py [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+#
+# http://numenta.org/licenses/
+# ----------------------------------------------------------------------
+
+from collections import OrderedDict
+
+import torch
+
+
+def upgrade_to_masked_sparseweights(state_dict):
```
Comment (lscheinkman) [inline nupic/torch/compatibility.py:27]: This function is not called here. Maybe you could use `_register_load_state_dict_pre_hook` in `SparseWeightsBase ` to replace the keys before `load_state_dict` is called

File: nupic/torch/modules/sparse_weights.py (hunk 1 above)
Comment (lscheinkman) [reply nupic/torch/modules/sparse_weights.py:127]: Yes, you could add `allow_extremes=False` to `SparseWeightsBase` but the leaving the assert on `SparseWeight` and  `SparseWeight2d` is a good compromise

File: nupic/torch/compatibility.py (hunk 3 above)
Comment (mrcslws) [reply nupic/torch/compatibility.py:27]: I played around with this. The problem is that it will require mutating the state dict that is passed in, which could cause issues. I also considered overriding `_load_from_state_dict`, but that would fail when `strict=False` (unless I mutate the state_dict).

`upgrade_to_masked_sparseweights` is called in nupic.research, by the way.
